WcaOnRails/app/models/competition.rb

```diff
@@ -193,13 +193,53 @@ class Competition < ApplicationRecord
 
   validates :registration_open, :registration_close, presence: { message: I18n.t('simple_form.required.text') }, if: :registration_period_required?
 
+  # NOTE: we only validate when confirming, until we have a unified events/rounds editor.
+  # If we would validate everytime, changing the number of rounds for
+  # competition wouldn't be possible: adding rounds through the events page
+  # couldn't be possible because the schedule doesn't contain the round
+  # just added.
   validate :must_have_at_least_one_event, if: :confirmed_or_visible?
   private def must_have_at_least_one_event
     if no_events?
       errors.add(:competition_events, I18n.t('competitions.errors.must_contain_event'))
     end
   end
 
+  # Only validate on update: nobody can confirm competition on creation.
+  # The only exception to this is within tests, in which case we actually don't want to run this validation.
+  validate :schedule_must_match_rounds, if: :confirmed_at_changed?, on: :update
+  def schedule_must_match_rounds
+    unless has_any_round_per_event? && schedule_includes_rounds?
+      errors.add(:competition_events, I18n.t('competitions.errors.schedule_must_match_rounds'))
+    end
+  end
+
+  def has_any_round_per_event?
+    competition_events.map { |ce| ce.rounds.any? }.reduce(&:&)
```

Comment:
Haha `&:&` looks cool, but what about `competition_events.map(&:rounds).none?(&:empty?)` :D

Reply:
Ah right, I didn't find `none?`, it's slightly more readable!